### PR TITLE
Add support for Python 3.x

### PR DESCRIPTION
**Pull request by nateskulic**

### CODE REVIEW

**Comment by rghe** [comment]
Hello, 
many thanks for your submission and please excuse the enormous delay in reviewing, I'm really sorry.
Anyway, the patch looks mostly ok, but here and there uses construct available only on python 3, like type hints.
While python 2 is now out of support I think lots of mercurial installation are still using python 2, so at this time perfarce should work with both.

File: perfarce.py [hunk 1]
```diff
@@ -220,15 +223,28 @@ def __del__(self):
         except Exception:
             pass
 
+def int_to_bytes(x: int) -> bytes:
```
Comment (rghe) [inline perfarce.py:226]: inline type informations are python 3 only.  A portable alternative would be using stub (.pyi) type files

File: perfarce.py [hunk 2]
```diff
@@ -366,31 +382,31 @@ def dothgonly(ctx):
         return node.nullid, 0
 
     @propertycache
-    def re_type(self): return re.compile('([a-z]+)?(text|binary|symlink|apple|resource|unicode|utf\d+)(\+\w+)?$')
+    def re_type(self): return re.compile(b'([a-z]+)?(text|binary|symlink|apple|resource|unicode|utf\d+)(\+\w+)?$')
     @propertycache
-    def re_keywords(self): return re.compile(r'\$(Id|Header|Date|DateTime|Change|File|Revision|Author):[^$\n]*\$')
+    def re_keywords(self): return re.compile(rb'\$(Id|Header|Date|DateTime|Change|File|Revision|Author):[^$\n]*\$')
```
Comment (rghe) [inline perfarce.py:387]: rb is python 3 only. br should work

File: perfarce.py [hunk 3]
```diff
@@ -479,42 +495,42 @@ def maxargs(self):
...
+            c.append(b'-d')
             c.append(shellquote(self.root))
 
         if files and len(files)>self.maxargs:
             tmp = TempFile('w')
             for f in files:
-                if self.ui.debugflag: self.ui.debug('> -x %s\n' % f)
-                print >>tmp.File,f
+                if self.ui.debugflag: self.ui.debug(b'> -x %s\n' % f)
+                print(f, file=tmp.File)
```
Comment (rghe) [inline perfarce.py:513]: python 3 only - from \_\_future\_\_ import print_function should work
or even tmp.File.write(f+'\n')

File: perfarce.py [hunk 4]
```diff
@@ -479,42 +495,42 @@ def maxargs(self):
...
-        cs = ' '.join(c + [shellquote(f) for f in files])
-        if self.ui.debugflag: self.ui.debug('> %s\n' % cs)
+        cs = b' '.join(c + [shellquote(f) for f in files])
+        if self.ui.debugflag: self.ui.debug(b'> %s\n' % cs)
 
-        for d in loaditer(os.popen(cs, 'rb')):
-            if self.ui.debugflag: self.ui.debug('< %r\n' % d)
-            code = d.get('code')
-            data = d.get('data')
+        for d in loaditer(popen(cs, b'rb')):
```
Comment (rghe) [inline perfarce.py:524]: loaditer expects a file argument. os.popen returns it, but mercurial.utils.procutil.popen does not

**Review by rghe — CHANGES_REQUESTED** [review]

**Comment by nateskulic** [comment]
Howdy!

Thanks for the comments.

Getting this to function required a patch to mercurial: https://www.mercurial-scm.org/repo/hg/rev/3af293735d0f

Tested against mercurial 5.3; and perforce 20.1. As of 5.3, our mercurial build has python3 as the default interpreter (shebang).

I also have a patch and fixes for the test-push-perforce script.

A few things:

- Mercurial should have dropped Python 2 support by now. "it is expected we will drop support for Python 2.7 sometime in 2020" [1]
- Python deprecated Python 2 as of Jan 2020. [2]
- I'd rather not waste another minute of life supporting python 2. Python 2 [workstation] audience includes no one I know.
- Github UI is a pain in the ass and working with/uploading patchwork here is also a pain in the ass --- so, I'd rather keep this process to absolute minimum.

Do you insist this must be fixed to python 2?

Cheers!


[1] https://www.mercurial-scm.org/wiki/Python3 
[2] https://www.python.org/doc/sunset-python-2/

**Comment by rghe** [comment]
Hi!
Well, mercurial still supports python 2, and will do so until at least 6.1 [1]
But even if mercurial and hg would drop py2 support tomorrow, not all users will be able to update at once.
I personally know of several users still needing python 2, especially on windows.
So I'd like perfarce to be compatibile with python 2 for a year at least, or until the burden becomes too heavy to bear :-)
This time the changes needed seem very small, so I will merge the patch and fix the compatibility issues myself.

[1] https://markmail.org/message/xn2wiarnb7jd2dlu

**Pull request merged**